File: 2594.minimum-time-to-repair-cars.ex
# @leet start
defmodule Solution do
  @spec repair_cars(ranks :: [integer], cars :: integer) :: integer
  def repair_cars(ranks, cars) do
    min_rank = Enum.min(ranks)
    left = 1
    right = min_rank * cars * cars
    binary_search(left, right, ranks, cars)
  end

  @spec can_repair?(ranks :: [integer], cars :: integer, time :: integer) :: boolean
  defp can_repair?(ranks, cars, time) do
    total_cars =
      Enum.reduce(ranks, 0, fn rank, acc ->
        (acc + :math.floor(:math.sqrt(time / rank))) |> trunc()
      end)

    total_cars >= cars
  end

  @spec binary_search(left :: integer, right :: integer, ranks :: [integer], cars :: integer) ::
          integer
  defp binary_search(left, right, _ranks, _cars) when left >= right(do: left)

  defp binary_search(left, right, ranks, cars) when left < right do
    mid = div(left + right, 2)

    if can_repair?(ranks, cars, mid) do
      binary_search(left, mid, ranks, cars)
    else
      binary_search(mid + 1, right, ranks, cars)
    end
  end
end

# @leet end

File: 2594.minimum-time-to-repair-cars.py
from typing import List


# @leet start
class Solution:
    def repairCars(self, ranks: List[int], cars: int) -> int:
        def can_repair_in_time(time: int) -> bool:
            total_cars = 0
            for rank in ranks:
                total_cars += int((time // rank) ** 0.5)
                if total_cars >= cars:
                    return True
            return False

        # actually, min_rank is a tighter upper bound
        # with min_rank (faster) repairing all cars by themself
        max_rank = max(ranks)
        lo = 1
        hi = max_rank * cars * cars + 1
        while lo < hi:
            mid = (lo + hi) // 2
            if can_repair_in_time(mid):
                hi = mid
            else:
                lo = mid + 1
        return lo


# @leet end

ranks = [1, 100]
cars = 2
print(Solution().repairCars(ranks, cars))
